rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+# RFC: Annotations
+
+_In which we propose the concept of and a draft API for ‘Annotations’,
+a new way for core and plugins to represent additional information about
+regions of a document._
+
+## Summary
+
+Add ‘annotations’, a new concept and API for the representation of
+transient document state. This provides a flexible mechanism for
+representing a wide variety of core and plugin information to the
+frontend, without using the text styling API.
+
+This API will be the basis for a wide variety of future features,
+including things like:
+
+- compiler warnings / lints / diagnostic messages
+- displaying other user's cursors during multi-user editing
+- clickable external links
+- version control status
+- breakpoints
```

Comment:
Are annotations the mechanism for conveying autocompletion suggestions to the front-end? I think so. In either case, it's good to be explicit. Obviously the annotation type is more complex than, say, a warning, so fleshing that out will be future work.

Reply:
I think that autocomplete is probably too stateful to really fit into this system, as I've been imagining it. Once sent, annotations should be valid until they are invalidated by an edit or by the provider; autocomplete wants to be invalidated by cursor movements, and are just in general far more transient; I think it needs its own API.

Reply:
Do they really have to be invalidated by cursor movement? The client could just hide them when they aren't relevant. And if there are edits they should indeed be invalidated anyway so the position of the cursor doesn't matter at that point.

Reply:
All I mean here is that if an autocomplete window is up and the user clicks elsewhere, autocomplete goes away; 'hiding' and 'invalidating' are equivalent under this model, since the client just draws what we send it.